File: app/src/main/java/com/pr0gramm/app/ApplicationClass.java
package com.pr0gramm.app;

import android.app.Application;
import android.content.Context;
import android.os.StrictMode;

import com.crashlytics.android.Crashlytics;
import com.f2prateek.dart.Dart;
import com.pr0gramm.app.services.ThemeHelper;
import com.pr0gramm.app.ui.ActivityErrorHandler;
import com.pr0gramm.app.util.CrashlyticsLogHandler;
import com.pr0gramm.app.util.Lazy;
import com.pr0gramm.app.util.LooperScheduler;
import com.thefinestartist.Base;

import net.danlew.android.joda.JodaTimeAndroid;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import butterknife.ButterKnife;
import io.fabric.sdk.android.Fabric;
import pl.brightinventions.slf4android.LoggerConfiguration;
import rx.Scheduler;
import rx.android.plugins.RxAndroidPlugins;
import rx.android.plugins.RxAndroidSchedulersHook;

import static com.pr0gramm.app.ui.dialogs.ErrorDialogFragment.setGlobalErrorDialogHandler;
import static com.pr0gramm.app.util.AndroidUtility.buildVersionCode;

/**
 * Global application class for pr0gramm app.
 */
public class ApplicationClass extends Application {
    private static final Logger logger = LoggerFactory.getLogger("Pr0grammApplication");

    final Lazy<AppComponent> appComponent = Lazy.of(() -> DaggerAppComponent.builder()
            .appModule(new AppModule(this))
            .httpModule(new HttpModule())
            .build());

    private static ApplicationClass INSTANCE;

    public ApplicationClass() {
        INSTANCE = this;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        Stats.init(buildVersionCode());
        JodaTimeAndroid.init(this);
        Base.initialize(this);

        Settings.initialize(this);

        if (BuildConfig.DEBUG) {
            logger.info("This is a development version.");
            StrictMode.enableDefaults();
            ButterKnife.setDebug(true);
            Dart.setDebug(true);

        } else {
            logger.info("Initialize fabric");
            Fabric.with(this, new Crashlytics());

            LoggerConfiguration.configuration()
                    .removeRootLogcatHandler()
                    .addHandlerToRootLogger(new CrashlyticsLogHandler());
        }

        // initialize this to show errors always in the context of the current activity.
        setGlobalErrorDialogHandler(new ActivityErrorHandler(this));

        // set as global constant that we can access from everywhere
        appComponent().googleAnalytics();

        Dagger.initEagerSingletons(this);

        if (BuildConfig.DEBUG) {
            StethoWrapper.init(this);
        }

        // get the correct theme for the app!
        ThemeHelper.updateTheme(this);
    }

    public static ApplicationClass get(Context context) {
        return (ApplicationClass) context.getApplicationContext();
    }

    public static AppComponent appComponent() {
        return INSTANCE.appComponent.get();
    }

    static {
        RxAndroidPlugins.getInstance().registerSchedulersHook(new RxAndroidSchedulersHook() {
            @Override
            public Scheduler getMainThreadScheduler() {
                return LooperScheduler.MAIN;
            }
        });
    }
}

File: app/src/main/java/com/pr0gramm/app/ApplicationClass.kt
package com.pr0gramm.app

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Bundle
import android.os.StrictMode
import android.util.Log
import androidx.work.Configuration
import androidx.work.WorkManager
import com.google.firebase.FirebaseApp
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.crashlytics.internal.common.CrashlyticsCore
import com.pr0gramm.app.services.ThemeHelper
import com.pr0gramm.app.services.Track
import com.pr0gramm.app.sync.SyncStatsWorker
import com.pr0gramm.app.sync.SyncWorker
import com.pr0gramm.app.ui.ActivityErrorHandler
import com.pr0gramm.app.ui.dialogs.ErrorDialogFragment.Companion.GlobalErrorDialogHandler
import com.pr0gramm.app.util.AndroidUtility
import com.pr0gramm.app.util.AndroidUtility.buildVersionCode
import com.pr0gramm.app.util.ExceptionHandler
import com.pr0gramm.app.util.debugOnly
import com.pr0gramm.app.util.di.InjectorAware
import com.pr0gramm.app.util.doInBackground
import kotlinx.coroutines.runBlocking
import java.util.WeakHashMap
import java.util.concurrent.TimeUnit
import java.util.logging.Level
import java.util.logging.LogManager

/**
 * Global application class for pr0gramm app.
 */
open class ApplicationClass : Application(), InjectorAware {
    private val bootupWatch = Stopwatch()

    private val logger = Logger("Pr0grammApp")

    init {
        debugOnly {
            System.setProperty("kotlinx.coroutines.debug", "on")
        }

        appContext = this
    }

    companion object {
        init {
            if (BuildConfig.DEBUG) {
                StrictMode.enableDefaults()

            } else {
                // allow all the dirty stuff.
                StrictMode.setVmPolicy(StrictMode.VmPolicy.LAX)
                StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.LAX)
            }
        }

        lateinit var appContext: Context
    }

    override fun onCreate() = logger.time("onCreate") {
        super.onCreate()

        val firebaseJob = doInBackground {
            logger.time("Initialize firebase") {
                initializeFirebase()

                // handler to ignore certain exceptions before they reach firebase
                ExceptionHandler.install(this)
            }
        }

        Stats.init(buildVersionCode())

        Settings.initialize(this)
        Track.initialize(this)

        logger.time("Initializing WorkManager") {
            WorkManager.initialize(this, Configuration.Builder()
                    .setMinimumLoggingLevel(if (BuildConfig.DEBUG) Log.VERBOSE else Log.INFO)
                    .build())
        }

        forceInjectorInstance()

        doInBackground {
            // schedule first sync 30seconds after bootup.
            SyncWorker.scheduleNextSyncIn(this@ApplicationClass, 30, TimeUnit.SECONDS, sourceTag = "Bootup")

            // also schedule the nightly update job
            SyncStatsWorker.schedule(this@ApplicationClass)
        }

        // initialize this to show errors always in the context of the current activity.
        GlobalErrorDialogHandler = ActivityErrorHandler(this)

        // get the correct theme for the app!
        ThemeHelper.updateTheme()

        if (!BuildConfig.DEBUG) {
            // disable verbose logging
            val log = LogManager.getLogManager().getLogger("")
            log?.handlers?.forEach { it.level = Level.INFO }
        }

        // wait for firebase setup to finish
        runBlocking {
            firebaseJob.join()
        }

        logger.info { "App booted in $bootupWatch" }

        Stats().histogram("app.boot.time", bootupWatch.elapsed().inMillis)
    }

    private fun initializeFirebase() {
        FirebaseApp.initializeApp(this)

        val fc = FirebaseCrashlytics.getInstance()

        val core = try {
            val field = FirebaseCrashlytics::class.java.declaredFields.first {
                it.type === CrashlyticsCore::class.java
            }

            field.isAccessible = true
            field.get(fc) as CrashlyticsCore

        } catch (err: Throwable) {
            debugOnly { throw err }
            null
        }

        if (core != null) {
            Logging.configureLoggingOutput { level, tag, message ->
                if (level >= Log.INFO) {
                    val levelStr = Logging.levels.getOrNull(level)
                    core.log("$levelStr [$tag]: $message")
                }
            }
        }
    }

    private val lifecycleCallbacksLookup = WeakHashMap<ActivityLifecycleCallbacks, CatchingActivityLifecycleCallbacks>()

    override fun registerActivityLifecycleCallbacks(callback: ActivityLifecycleCallbacks) {
        // Issue with crashlytics / measurement. onActivityCreate throws a 'ConcurrentModificationException'
        // that then crashes the instance. This way we can catch & ignore issues directly.
        val wrapped = CatchingActivityLifecycleCallbacks(callback)
        lifecycleCallbacksLookup[callback] = wrapped
        super.registerActivityLifecycleCallbacks(wrapped)
    }

    override fun unregisterActivityLifecycleCallbacks(callback: ActivityLifecycleCallbacks) {
        super.unregisterActivityLifecycleCallbacks(lifecycleCallbacksLookup[callback] ?: callback)
    }

    private fun forceInjectorInstance() {
        // ensure that the lazy creates the instance
        System.identityHashCode(injector)

        debugOnly {
            // validate that all dependencies can be created.
            injector.validate()
        }
    }

    override val injector by lazy { appInjector(this) }
}


private class CatchingActivityLifecycleCallbacks(
        private val callback: Application.ActivityLifecycleCallbacks)
    : Application.ActivityLifecycleCallbacks by callback {

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        try {
            callback.onActivityCreated(activity, savedInstanceState)
        } catch (err: RuntimeException) {
            AndroidUtility.logToCrashlytics(IllegalStateException(
                    "ActivityLifecycleCallback failed with ${err.javaClass.simpleName}",
                    err))
        }
    }
}
